src/main/scala/gemmini/ReservationStation.scala

```diff
@@ -281,91 +269,77 @@ class ReservationStation[T <: Data : Arithmetic, U <: Data, V <: Data](config: G
-      m && op1.valid && e.bits.q =/= new_entry.q && e.bits.opa_is_dst
-    })
-    val op2_raws_opa = VecInit((entries zip op2_matches_opa).map { case (e, m) =>
-      m && op2.valid && e.bits.q =/= new_entry.q && e.bits.opa_is_dst
-    })
-    val raws = VecInit((op1_raws_opa zip op2_raws_opa).map { case (a, b) => a || b })
-
-    val dst_wars_opa = VecInit((entries zip dst_matches_opa).map { case (e, m) =>
-      m && dst.valid && e.bits.q =/= new_entry.q && !e.bits.opa_is_dst
-    })
-    val dst_wars_opb = VecInit((entries zip dst_matches_opb).map { case (e, m) =>
-      m && dst.valid && e.bits.q =/= new_entry.q
-    })
-    val wars = VecInit((dst_wars_opa zip dst_wars_opb).map { case (a, b) => a || b })
-
-    val dst_waws_opa = VecInit((entries zip dst_matches_opa).map { case (e, m) =>
-      m && dst.valid && (e.bits.q =/= new_entry.q || new_entry.q === ldq) && e.bits.opa_is_dst
-    })
-    val waws = dst_waws_opa
-
-    val older_in_same_q = VecInit(entries.map { e =>
-      e.valid && e.bits.q === new_entry.q && !e.bits.issued
-    })
-
-    val is_st_and_must_wait_for_prior_ex_config = VecInit(entries.map { e =>
-      e.valid && new_entry.q === stq && !new_entry.is_config && e.bits.q === exq && e.bits.is_config
-    })
-
-    val is_ex_config_and_must_wait_for_prior_st = VecInit(entries.map { e =>
-      // TODO when acc reads no longer rely upon config-ex's relu6, this dependency can be broken
-      e.valid && new_entry.q === exq && new_entry.is_config && e.bits.q === stq && !e.bits.is_config
-    })
-
-    new_entry.deps := (Cat(raws) | Cat(wars) | Cat(waws) | Cat(older_in_same_q) |
-      Cat(is_st_and_must_wait_for_prior_ex_config) | Cat(is_ex_config_and_must_wait_for_prior_st)).asBools().reverse
-
-    raws_probe := Cat(raws.reverse)
-    waws_probe := Cat(waws.reverse)
-    wars_probe := Cat(wars.reverse)
-    older_in_same_q_probe := Cat(older_in_same_q.reverse)
-    is_st_and_must_wait_for_prior_ex_config_probe := Cat(is_st_and_must_wait_for_prior_ex_config.reverse)
-    is_ex_config_and_must_wait_for_prior_st_probe := Cat(is_ex_config_and_must_wait_for_prior_st.reverse)
+    assert(ldq < exq && exq < stq)
+
+    val not_config = !new_entry.is_config
+    when (is_load) {
+      // war (after ex/st) | waw (after ex)
+      val ld_deps = VecInit(entries_ld.map { e => e.valid && !e.bits.issued }) // same q - can this be ignored?
+
+      val ex_deps = VecInit(entries_ex.map { e => e.valid && !new_entry.is_config && (
+        (new_entry.opa.bits.overlaps(e.bits.opa.bits) && e.bits.opa.valid) || // waw if preload, war if compute
+        (new_entry.opa.bits.overlaps(e.bits.opb.bits) && e.bits.opb.valid))}) // war
+
+      val st_deps = VecInit(entries_st.map { e => e.valid && e.bits.opa.valid && not_config &&
+        new_entry.opa.bits.overlaps(e.bits.opa.bits)})  // war
+
+      new_entry.deps := ld_deps ++ ex_deps ++ st_deps // TODO: doesn't work when entries per type isnt power of 2
```

Comment:
I'm perfectly fine with restricting entries-per-type to be a power of 2.

But you need to add something like an `assert(isPow2(ld_deps.size))` if you're making that assumption

Reply:
Addressed at the start of the reservation station; each entry count is tested